markdown/extensions/toc.py

```diff
@@ -136,6 +136,11 @@ def __init__(self, md, config):
             self.use_permalinks = config["permalink"]
         self.header_rgx = re.compile("[Hh][123456]")
         self.toc_depth = config["toc_depth"]
+        if type(self.toc_depth) is not int:
```

Comment:
First of all, we don't need to assign `self.toc_depth` as we don't need it anymore. `self.toc_top` and self.toc_bottom` serve that purpose just fine. In the next few lines we can just reference `config["toc_depth"]`.

Sorry I missed this last time, but `if type(foo) is bar` is not the correct way to do a type test in Python. You should use `if is_instance(foo, bar)`, which allows users to pass in subclasses of the specified type (in other words, support duck-typing).

Thirdly, this does not account for `self.toc_depth` being a single integer within a string (`"2"`). 

There are a couple ways we could address this:

1. We could do what I suggested the first time:

    ```python
    try:
         self.toc_bottom = int(config["toc_depth"])
         self.toc_top = 1
    except ValueError:
         self.toc_top, self.toc_bottom = [int(t), int(b) for t, b in config["toc_depth"].split('-')]
    ```

2. Or we could check for a hyphen:

    ```python
    if is_instance(config["toc_depth"], base_string) and '-' in config["toc_depth"]:
        self.toc_top, self.toc_bottom = [int(t), int(b) for t, b in config["toc_depth"].split('-')]
    else:
        self.toc_top = 1
        self.toc_bottom = int(config["toc_depth"])

Notice that on both solutions, we explicitly force all values to integers. Therefore, we do not need to do that later when we use the values.

As it turns out, I like the second solution better as it keeps the old behavior and simply does a conditional check for the new syntax. Any bad data will still act exactly the same as it would have before the change.

Reply:
I'd make these changes myself, but you used your `master` branch which makes it difficult for me to push any changes to. If you apply the following diff exactly as-is, I'll merge this PR:

```diff
--- a/markdown/extensions/toc.py
+++ b/markdown/extensions/toc.py
@@ -135,12 +135,11 @@ class TocTreeprocessor(Treeprocessor):
         if self.use_permalinks is None:
             self.use_permalinks = config["permalink"]
         self.header_rgx = re.compile("[Hh][123456]")
-        self.toc_depth = config["toc_depth"]
-        if type(self.toc_depth) is not int:
-            self.toc_top, self.toc_bottom = self.toc_depth.split('-')
+        if isinstance(config["toc_depth"], basestring) and '-' in config["toc_depth"]:
+            self.toc_top, self.toc_bottom = [int(x) for x in config["toc_depth"].split('-')]
         else:
             self.toc_top = 1
-            self.toc_bottom = self.toc_depth
+            self.toc_bottom = int(config["toc_depth"])
 
     def iterparent(self, node):
         ''' Iterator wrapper to get allowed parent and child all at once. '''
@@ -240,7 +239,7 @@ class TocTreeprocessor(Treeprocessor):
         for el in doc.iter():
             if isinstance(el.tag, string_type) and self.header_rgx.match(el.tag):
                 self.set_level(el)
-                if int(el.tag[-1]) < int(self.toc_top) or int(el.tag[-1]) > int(self.toc_bottom):
+                if int(el.tag[-1]) < self.toc_top or int(el.tag[-1]) > self.toc_bottom:
                     continue
                 text = ''.join(el.itertext()).strip()
```

Reply:
@waylan thanks for your help. I used your suggestions, but it failed on

```suggestion
        if isinstance(config["toc_depth"], basestring) and '-' in config["toc_depth"]:
```
 with: _NameError: name 'basestring' is not defined_

'basestring' is python2 only, can I use [string_types from six](https://pythonhosted.org/six/#six.string_types) instead?

PS: I don't want to push non working code, so I will wait for you.

Reply:
Sorry, I forgot about that. six is why overkill for what we need. You can use `markdown.util.string_type`, which is already imported by the module as `string_type`. So just replace `basestring` with `string_type`.

Reply:
@waylan Oh i missed looking into markdown/util.py, this is great;) I committed your suggestions.